src/templates/k8soperator/Content/Operator.cs

```diff
@@ -0,0 +1,138 @@
+﻿using __ProjectName__.Controller;
+using __ProjectName__.Crd;
+using __ProjectName__.Diagnostics;
+using k8s;
+using Microsoft.Extensions.Hosting;
+using System;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace __ProjectName__
+{
+    public class Operator : IHostedService
+    {
+        private Watcher<{{crdname}}Resource> _watcher;
+
+        private readonly Channel<WatchData> _channel;
+
+        private readonly CancellationTokenSource _stoppingCts = new CancellationTokenSource();
+        private readonly IKubernetes _kubernetesClient;
+        private readonly I__ProjectName__Controller _controller;
+        private readonly OperatorDiagnostics _diagnostics;
+
+        public Operator(
+            IKubernetes kubernetesClient,
+            I__ProjectName__Controller controller,
+            OperatorDiagnostics diagnostics)
+        {
+            _kubernetesClient = kubernetesClient ?? throw new ArgumentNullException(nameof(kubernetesClient));
+            _controller = controller ?? throw new ArgumentNullException(nameof(controller));
+            _diagnostics = diagnostics ?? throw new ArgumentNullException(nameof(diagnostics));
+
+            _channel = Channel.CreateUnbounded<WatchData>(new UnboundedChannelOptions()
+            {
+                SingleReader = true,
+                SingleWriter = true
+            });
+        }
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            _diagnostics.OperatorStarting();
+
+            //starting the channel reader and the watcher, we are not awaiting the watcher
+            // because this is on wait state meanwhile not crd objects are present/created 
+            //on K8S and the hosted initialization can't be continue
+
+            Task.Run(WatchListener);
+            _ = StartWatcher(_stoppingCts.Token);
+
+            return Task.CompletedTask;
+        }
+
+        public Task StopAsync(CancellationToken cancellationToken)
+        {
+            _diagnostics.OperatorShuttingdown();
+
+            _stoppingCts.Cancel();
+            _channel.Writer.Complete();
+            
+            if ( _watcher != null )
+            {
+                //if response is not finished when hosted service is 
+                //on close this object will be null.
+
+                _watcher.Dispose();
+            }
+            
+            return Task.CompletedTask;
+        }
+      
+        private async Task StartWatcher(CancellationToken cancellationToken)
+        {
+            var response = await _kubernetesClient.ListClusterCustomObjectWithHttpMessagesAsync(

```

Comment:
What do you want to happen if this fails??

Reply:
We should implement some retry strategy for transient errors?

Reply:
Do you want the app to fail startup or retry?

Reply:
I tested again the watcher and if any exception is throw  this perform the onClose again, the onclose code do retries starting the watcher again.

```csharp
onClosed: ()
{
        // watcher is closed and reactivated automatically 
        // every "timeoutSeconds"  or when any exception is thrown
        if (!_stoppingCts.IsCancellationRequested)
        {
              _watcher.Dispose();
              _ = StartWatcher(_stoppingCts.Token);
        }
},
```